File: sagemaker-pipelines/tabular/customizing_build_train_deploy_project/modelbuild/pipelines/customer_churn/pipeline.py
# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://aws.amazon.com/apache2.0/
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.
"""Example workflow pipeline script for CustomerChurn pipeline.

                                               . -RegisterModel
                                              .
    Process-> Train -> Evaluate -> Condition .
                                              .
                                               . -(stop)

Implements a get_pipeline(**kwargs) method.
"""

import os

import boto3
import sagemaker
import sagemaker.session

from sagemaker.estimator import Estimator
from sagemaker.inputs import TrainingInput
from sagemaker.processing import (
    ProcessingInput,
    ProcessingOutput,
    ScriptProcessor,
)
from sagemaker.sklearn.processing import SKLearnProcessor
from sagemaker.workflow.conditions import (
    ConditionGreaterThanOrEqualTo,
)
from sagemaker.workflow.condition_step import (
    ConditionStep,
    JsonGet,
)
from sagemaker.model_metrics import (
    MetricsSource,
    ModelMetrics,
)
from sagemaker.workflow.parameters import (
    ParameterInteger,
    ParameterString,
)
from sagemaker.workflow.pipeline import Pipeline
from sagemaker.workflow.properties import PropertyFile
from sagemaker.workflow.steps import (
    ProcessingStep,
    TrainingStep,
)
from sagemaker.workflow.step_collections import RegisterModel


BASE_DIR = os.path.dirname(os.path.realpath(__file__))


def get_session(region, default_bucket):
    """Gets the sagemaker session based on the region.

    Args:
        region: the aws region to start the session
        default_bucket: the bucket to use for storing the artifacts

    Returns:
        `sagemaker.session.Session instance
    """

    boto_session = boto3.Session(region_name=region)

    sagemaker_client = boto_session.client("sagemaker")
    runtime_client = boto_session.client("sagemaker-runtime")
    return sagemaker.session.Session(
        boto_session=boto_session,
        sagemaker_client=sagemaker_client,
        sagemaker_runtime_client=runtime_client,
        default_bucket=default_bucket,
    )


def get_pipeline(
    region,
    role=None,
    default_bucket=None,
    model_package_group_name="CustomerChurnPackageGroup",  # Choose any name
    pipeline_name="CustomerChurnDemo-p-ewf8t7lvhivm",  # You can find your pipeline name in the Studio UI (project -> Pipelines -> name)
    base_job_prefix="CustomerChurn",  # Choose any name
):
    """Gets a SageMaker ML Pipeline instance working with on CustomerChurn data.

    Args:
        region: AWS region to create and run the pipeline.
        role: IAM role to create and run steps and pipeline.
        default_bucket: the bucket to use for storing the artifacts

    Returns:
        an instance of a pipeline
    """
    sagemaker_session = get_session(region, default_bucket)
    if role is None:
        role = sagemaker.session.get_execution_role(sagemaker_session)

    # Parameters for pipeline execution
    processing_instance_count = ParameterInteger(
        name="ProcessingInstanceCount", default_value=1
    )
    processing_instance_type = ParameterString(
        name="ProcessingInstanceType", default_value="ml.m5.xlarge"
    )
    training_instance_type = ParameterString(
        name="TrainingInstanceType", default_value="ml.m5.xlarge"
    )
    model_approval_status = ParameterString(
        name="ModelApprovalStatus",
        default_value="PendingManualApproval",  # ModelApprovalStatus can be set to a default of "Approved" if you don't want manual approval.
    )
    input_data = ParameterString(
        name="InputDataUrl",
        default_value=f"s3://sm-pipelines-demo-data-123456789/churn.txt",  # Change this to point to the s3 location of your raw input data.
    )

    # Processing step for feature engineering
    sklearn_processor = SKLearnProcessor(
        framework_version="0.23-1",
        instance_type=processing_instance_type,
        instance_count=processing_instance_count,
        base_job_name=f"{base_job_prefix}/sklearn-CustomerChurn-preprocess",  # choose any name
        sagemaker_session=sagemaker_session,
        role=role,
    )
    step_process = ProcessingStep(
        name="CustomerChurnProcess",  # choose any name
        processor=sklearn_processor,
        outputs=[
            ProcessingOutput(output_name="train", source="/opt/ml/processing/train"),
            ProcessingOutput(
                output_name="validation", source="/opt/ml/processing/validation"
            ),
            ProcessingOutput(output_name="test", source="/opt/ml/processing/test"),
        ],
        code=os.path.join(BASE_DIR, "preprocess.py"),
        job_arguments=["--input-data", input_data],
    )

    # Training step for generating model artifacts
    model_path = f"s3://{sagemaker_session.default_bucket()}/{base_job_prefix}/CustomerChurnTrain"
    image_uri = sagemaker.image_uris.retrieve(
        framework="xgboost",  # we are using the Sagemaker built in xgboost algorithm
        region=region,
        version="1.0-1",
        py_version="py3",
        instance_type=training_instance_type,
    )
    xgb_train = Estimator(
        image_uri=image_uri,
        instance_type=training_instance_type,
        instance_count=1,
        output_path=model_path,
        base_job_name=f"{base_job_prefix}/CustomerChurn-train",
        sagemaker_session=sagemaker_session,
        role=role,
    )
    xgb_train.set_hyperparameters(
        objective="binary:logistic",
        num_round=50,
        max_depth=5,
        eta=0.2,
        gamma=4,
        min_child_weight=6,
        subsample=0.7,
        silent=0,
    )
    step_train = TrainingStep(
        name="CustomerChurnTrain",
        estimator=xgb_train,
        inputs={
            "train": TrainingInput(
                s3_data=step_process.properties.ProcessingOutputConfig.Outputs[
                    "train"
                ].S3Output.S3Uri,
                content_type="text/csv",
            ),
            "validation": TrainingInput(
                s3_data=step_process.properties.ProcessingOutputConfig.Outputs[
                    "validation"
                ].S3Output.S3Uri,
                content_type="text/csv",
            ),
        },
    )

    # Processing step for evaluation
    script_eval = ScriptProcessor(
        image_uri=image_uri,
        command=["python3"],
        instance_type=processing_instance_type,
        instance_count=1,
        base_job_name=f"{base_job_prefix}/script-CustomerChurn-eval",
        sagemaker_session=sagemaker_session,
        role=role,
    )
    evaluation_report = PropertyFile(
        name="EvaluationReport",
        output_name="evaluation",
        path="evaluation.json",
    )
    step_eval = ProcessingStep(
        name="CustomerChurnEval",
        processor=script_eval,
        inputs=[
            ProcessingInput(
                source=step_train.properties.ModelArtifacts.S3ModelArtifacts,
                destination="/opt/ml/processing/model",
            ),
            ProcessingInput(
                source=step_process.properties.ProcessingOutputConfig.Outputs[
                    "test"
                ].S3Output.S3Uri,
                destination="/opt/ml/processing/test",
            ),
        ],
        outputs=[
            ProcessingOutput(
                output_name="evaluation", source="/opt/ml/processing/evaluation"
            ),
        ],
        code=os.path.join(BASE_DIR, "evaluate.py"),
        property_files=[evaluation_report],
    )

    # Register model step that will be conditionally executed
    model_metrics = ModelMetrics(
        model_statistics=MetricsSource(
            s3_uri="{}/evaluation.json".format(
                step_eval.arguments["ProcessingOutputConfig"]["Outputs"][0]["S3Output"][
                    "S3Uri"
                ]
            ),
            content_type="application/json",
        )
    )

    # Register model step that will be conditionally executed
    step_register = RegisterModel(
        name="CustomerChurnRegisterModel",
        estimator=xgb_train,
        model_data=step_train.properties.ModelArtifacts.S3ModelArtifacts,
        content_types=["text/csv"],
        response_types=["text/csv"],
        inference_instances=["ml.t2.medium", "ml.m5.large"],
        transform_instances=["ml.m5.large"],
        model_package_group_name=model_package_group_name,
        approval_status=model_approval_status,
        model_metrics=model_metrics,
    )

    # Condition step for evaluating model quality and branching execution
    cond_lte = ConditionGreaterThanOrEqualTo(  # You can change the condition here
        left=JsonGet(
            step=step_eval,
            property_file=evaluation_report,
            json_path="binary_classification_metrics.accuracy.value",  # This should follow the structure of your report_dict defined in the evaluate.py file.
        ),
        right=0.8,  # You can change the threshold here
    )
    step_cond = ConditionStep(
        name="CustomerChurnAccuracyCond",
        conditions=[cond_lte],
        if_steps=[step_register],
        else_steps=[],
    )

    # Pipeline instance
    pipeline = Pipeline(
        name=pipeline_name,
        parameters=[
            processing_instance_type,
            processing_instance_count,
            training_instance_type,
            model_approval_status,
            input_data,
        ],
        steps=[step_process, step_train, step_eval, step_cond],
        sagemaker_session=sagemaker_session,
    )
    return pipeline

File: r_examples/rstudio_sagemaker_pipelines_and_serving/pipeline.R
library(reticulate)

local_path <- dirname(rstudioapi::getSourceEditorContext()$path)

get_pipeline <- function(input_data_uri){
  
  s3_raw_data <- sagemaker$workflow$parameters$ParameterString(
    name="InputData",
    default_value=input_data_uri
  )
  
  sagemaker <- import('sagemaker')
  boto3 <- import('boto3')
  
  session <- sagemaker$Session()
  bucket <- session$default_bucket()
  
  role_arn <- sagemaker$get_execution_role()
  account_id <- boto3$client("sts")$get_caller_identity()$"Account"
  region <- boto3$session$Session()$region_name
  
  ecr_repository_processing = "sagemaker-r-processing"
  tag_processing <- ":1.0"
  
  ecr_repository_training = "sagemaker-r-train-and-deploy"
  tag_training <- ":1.0"
  
  
  # Define preprocessing step  ------------------
  processing_repository_uri <- paste0(account_id,".dkr.ecr.",region,".amazonaws.com","/",ecr_repository_processing,tag_processing)
  
  
  script_processor <- sagemaker$processing$ScriptProcessor(command=list("Rscript"), 
                                                           image_uri=processing_repository_uri,
                                                           role=role_arn,
                                                           instance_count=1L,
                                                           instance_type="ml.m5.large")
  
  output_s3_data_location <- paste0("s3://", bucket, "/pipeline-example/processing_output")
  s3_processing_input <- sagemaker$processing$ProcessingInput(source = s3_raw_data, destination="/opt/ml/processing/input")
  s3_processing_output1 <- sagemaker$processing$ProcessingOutput(output_name="abalone_train", destination=output_s3_data_location, source="/opt/ml/processing/output/train")
  s3_processing_output2 <- sagemaker$processing$ProcessingOutput(output_name="abalone_test",  destination=output_s3_data_location, source="/opt/ml/processing/output/test")
  s3_processing_output3 <- sagemaker$processing$ProcessingOutput(output_name="abalone_valid", destination=output_s3_data_location, source="/opt/ml/processing/output/valid")
  
  step_process <- sagemaker$workflow$steps$ProcessingStep(name="ProcessingStep",
                                                          code=paste0(local_path, "/preprocessing/preprocessing.R"),
                                                          processor=script_processor,
                                                          inputs=list(s3_processing_input),
                                                          outputs=list(s3_processing_output1, s3_processing_output2, s3_processing_output3))
  
  
  # Define training step  ------------------
  model_image_uri <- paste0(account_id,".dkr.ecr.",region,".amazonaws.com","/",ecr_repository_training,tag_training)
  
  train_estimator <- sagemaker$estimator$Estimator(model_image_uri,
                                                   role_arn, 
                                                   source_dir=paste0(local_path,"/training/code/"),
                                                   entry_point="train.R",
                                                   instance_count=1L, 
                                                   instance_type='ml.m5.xlarge',
                                                   output_path=paste0("s3://", bucket, "/pipeline-example/training_output"),
                                                   sagemaker_session=session,
                                                   metric_definitions=list(list(Name="rmse-validation", Regex= "Calculated validation RMSE: ([0-9.]+);.*$")))
  
  step_train <- sagemaker$workflow$steps$TrainingStep(
    name="TrainingStep",
    estimator=train_estimator,
    inputs=list(training = sagemaker$inputs$TrainingInput(
                            s3_data=step_process$properties$ProcessingOutputConfig$Outputs["abalone_train"]$S3Output$S3Uri,
                            content_type="text/csv")
                )
  )
  
  # Evaluate model step -----------------
  
  script_eval <- sagemaker$processing$ScriptProcessor(
    image_uri=processing_repository_uri,
    command=list("Rscript"),
    instance_type="ml.m5.large",
    instance_count=1L,
    base_job_name="script-evaluate",
    role=role_arn
  )
  
  evaluation_report <- sagemaker$workflow$properties$PropertyFile(
    name="EvaluationReport", output_name="evaluation", path="evaluation.json"
  )
  
  step_eval = sagemaker$workflow$steps$ProcessingStep(
    name="EvaluateModel",
    processor=script_eval,
    inputs=list(
      sagemaker$processing$ProcessingInput(
        source=step_train$properties$ModelArtifacts$S3ModelArtifacts,
        destination="/opt/ml/processing/model"
      ),
      sagemaker$processing$ProcessingInput(
        source=step_process$properties$ProcessingOutputConfig$Outputs["abalone_test"]$S3Output$S3Uri,
        destination="/opt/ml/processing/test"
      )
    ),
    outputs=list(
      sagemaker$processing$ProcessingOutput(output_name="evaluation", source="/opt/ml/processing/evaluation")
    ),
    code=paste0(local_path, "/postprocessing/evaluation.R"),
    property_files=list(evaluation_report)
  )
  
  
  # Register model ------------------
  
  model_metrics <- sagemaker$model_metrics$ModelMetrics(
    model_statistics=sagemaker$model_metrics$MetricsSource(
      s3_uri=paste0(step_eval$arguments[["ProcessingOutputConfig"]][["Outputs"]][[1]][["S3Output"]][["S3Uri"]], "/evaluation.json"),
      content_type="application/json"
    )
  )
  
  step_register <- sagemaker$workflow$step_collections$RegisterModel(
    name="RegisterModelStep",
    estimator=train_estimator,
    model_data=step_train$properties$ModelArtifacts$S3ModelArtifacts,
    content_types=list("application/json"),
    response_types=list("application/json"),
    inference_instances=list("ml.t2.medium", "ml.m5.xlarge"),
    transform_instances=list("ml.m5.xlarge"),
    model_package_group_name="AbaloneRModelPackageGroup",
    approval_status="Approved",  # we are automatically registering the model as approved but in more general case you'd want to set this to "PendingManualApproval",
    model_metrics=model_metrics
  )
  
  # Condition Step
  
  cond_lte <- sagemaker$workflow$conditions$ConditionLessThanOrEqualTo(
    left=sagemaker$workflow$functions$JsonGet(
      step_name=step_eval$name,
      property_file=evaluation_report,
      json_path="regression_metrics.rmse.value"
    ),
    right=6.0
  )
  
  step_cond <- sagemaker$workflow$condition_step$ConditionStep(
    name="rmseConditional",
    conditions=list(cond_lte),
    if_steps=list(step_register),
    else_steps=list()
  )
  
  
  # Construct the Pipeline
  
  pipeline_name <-"AbalonePipelineUsingR"
  pipeline = sagemaker$workflow$pipeline$Pipeline(
    name=pipeline_name,
    parameters=list(s3_raw_data),
    steps=list(step_process, step_train, step_eval, step_cond)
  )
  
  return(pipeline)
}